docs: improve search for documentation

tmair opened the pull request:
<!--
Thank you very much for your pull request!

If your PR is the addition of a new operator, please make sure all these boxes are ticked with an x:

- [ ] Add the operator to Rx
- [ ] It must have a `-spec.ts` tests file covering the canonical corner cases, with marble diagram tests
- [ ] The spec file should have a type definition test at the end of the spec to verify type definition for various use cases
- [ ] The operator must be documented in JSDoc style in the implementation file, including also the PNG marble diagram image
- [ ] The operator should be listed in `docs_app/content/guide/operators.md` in a category of operators
- [ ] The operator should also be documented. See [Documentation Guidelines](../CONTRIBUTING.md).
- [ ] You may need to update `MIGRATION.md` if the operator differs from the corresponding one in RxJS v4
-->

**Description:**
This PR ports the search web worker and its integration from angular.io to the RxJS documentation site. The main changes are outlined on the original commit in the angular repository: https://github.com/angular/angular/commit/fccffc647b6b4ee2a4e15b6d3563e24f853fb0d5

In Addition to the changes ported from the angular documentation the following changes where made:
- Adaption of the default lunr stop word filter to exclude `from`, `of` and `every` as a stop word (this allows for a search of the `of` function)
- ~~disabling of the ignore word list for class and interface members (this allows to find results for `next` as `next` is a member of the `Observable` interface)~~ This is now also fixed on angular.io

**Note:**
This PR needs to be rebased after the dependency issues in #6913 are merged.

<!--
If this PR introduces a breaking change, it must contain a notice for it to be included in the CHANGELOG.
-->


**Related issue (if exists):**
This should fix https://github.com/ReactiveX/rxjs/issues/6500
The should be no regression for https://github.com/reactivex/rxjs/issues/4536

jakovljevic-mladen reviewed (commented):
This is a great PR, thank you a lot for making it. It took me a while until I got into this stuff so I could do the proper review. Changes look good to me, however, I've got some questions/comments.

One of them is related to the duplicated results (not related to this PR, but could be handy if we fixed the issue). For example, in this picture:

![image](https://user-images.githubusercontent.com/28087049/178280613-9d95f9e4-1340-4c7c-b5bc-48066d535272.png)

You can see that we have `repeatWhen` appearing two times (the on is the link to this path: `/api/index/function/repeatWhen`, and the other one to this path: `/api/operators/repeatWhen`). Is it possible to remove docs with `/api/operators/` path (since this path is deprecated)?

Other questions/comments in files.

tmair commented:
Hi @jakovljevic-mladen,

rhabks for the review. I will look at your questions but right now I don't have the time for it. It will take me 3 to 4 weeks to get back to it, but I will continue the work on it

tmair commented on docs_app/tools/transforms/angular-base-package/processors/generateKeywords.js:
```diff
@@ -1,7 +1,5 @@
-
-var fs = require('fs');
-var path = require('canonical-path');
+/* eslint-disable */
```
There is a precommit hook doing an eslint check. I can push using `--no-verify` if you would like to keep those comments out of the code base.

jakovljevic-mladen replied:
I didn't actually know that we have ESLint rules enabled 😄 I guess the line can stay in that case.

tmair commented on docs_app/.gitignore:
```diff
@@ -47,7 +47,6 @@ protractor-results*.txt
 Thumbs.db
 
 # copied dependencies
```
Yes. I removed that line

tmair commented on docs_app/tools/transforms/angular-base-package/ignore-words.json:
```diff
@@ -0,0 +1,705 @@
+{
+  "en": [
```
I changed the word list to only an array.

tmair commented on docs_app/src/app/search/search.worker.ts:
```diff
@@ -0,0 +1,276 @@
+  'we',
+  'were',
+  'what',
+  'when',
```
Stop words will match only the stop word. So when the word `when` is included within a term that term is not filtered.

tmair commented:
I rebased the branch on the current master and tried to incorporate your review feedback into my branch. 

Since the search is somewhat difficult to understand I have made some notes on how the search works. I hope this helps to understand the code base better and should also serve as a reference for future work in this area.

# How does the search work?

The search is split into a preprocessing step at build time and a index creation step (inside a web worker) on runtime.

The preprocessing happens in `generateKeywords.js` and performs the following steps:

1. calculate the set of pages that needs to be indexed for searching
2. For each non ignored property (`ignoredProperties`) of each page document, extract keywords used to build the search index (`mainTokens`). Here the ignore words list is used to filter out common english terminology (e.g. you, from, every, ... )
3. Extract all public members from the code and add those to the search index (`memberTokens`) ignoring the ignore words list.
4. Extract keywords from the headings of the documents (`headingTokens`) using the ignore word list
5. Generate the search data that will be downloaded by the service worker to build the search index. The search data is already stemmed (https://lunrjs.com/guides/core_concepts.html#stemming) to further reduce the amount of data needed to transfer to the service worker.

The service worker performs the following steps, before it can be used to answer any search results:

1. Load the search data computed by the preprocessing step described above
2. Create a processing pipeline by not including a stemming step, as this was already performed during preprocessing. However a stopWord processing step is performed.
3. Process the data through the processing pipeline to create a search index


During the preprocessing and indexing of the search data there are two places where words are ignored. The ignore word list, will be used to filter common english terms during the preprocessing phase (not used with member tokens). The stop word filter will perform the same operation when building the index on the service worker.

# Improvements to the searching process done in this PR

Because it is strange to have two ignore word lists and filtering steps, I have decided to remove the stop word filter from the index builder pipeline completly, as it is does the work twice and is quite confusing. However I have left all words (including from, every, of, with) included in the ignore word filter, as the search results are not great when those words are not ignored. For example `from` is used whenever there is an example included within a documentation page because of the `import ... from ...` syntax. The relevant pages are still found, because the title of the page is not filtered. Therefore `from` will still return the search results where `from` is included within the title of the page.

Regarding the different search results for from, here is what is happening during the search:
1. First search for a document that contains all of the search terms (divided by spaces) provided (in our case `from`)
2. If the search has at least one result, return that result set (and that is what is happening for `from`. Thus we only have one search result)
3. Search for a document that contains any of the search terms
4. If the search has at least one result, return that result set 
5. Search for documents that contain any of the search terms and contain the first search term somewhere in the tile

So for `from` as there is only one result, we return that single result. 

I modified the initial search in such a way, that it will add more weight to search terms within the title of the page. Maybe that will result in better search results.

tmair commented on docs_app/tools/transforms/angular-base-package/ignore-words.json:
```diff
@@ -0,0 +1,705 @@
+    "forth",
+    "found",
+    "four",
+    "from",
```
Please see the explanation of this issue in my comment below.

benlesh commented:
@jakovljevic-mladen is this still something you're considering merging?

jakovljevic-mladen commented:
@benlesh, yes, but I'm still waiting for #6913 to be merged. I should've wrote it in the comments so that @tmair knows as well, I'm so sorry about that.

benlesh commented:
@jakovljevic-mladen #6913 is approved and ready to merge. Sorry for the delay.

jakovljevic-mladen reviewed (approved):
@tmair, thank you a lot for making these changes. I read your exhaustive explanation of how search works and I agree with these changes. During my testing phase couple of months ago, I had issues when I was searching term `from` where my results were different to what you were describing. So, I assumed this is due to conflicting dependencies. Now that #6913 is merged, I rebased this PR against master, did `npm install` and tried search and it worked the way you described it.

Since I did rebase, I'd like you to review this PR to check if merge conflicts are resolved correctly. Also, we have issues with merging to master due to issue with Node 14 which is now removed, but pipelines still show this as a required step.

tmair commented:
@jakovljevic-mladen I reviewed the PR and also did a short local test of the new search implementation. So far everything looks good.

The only thing that was different for me was a `npm install` will alter the `package-lock.json`. I don't know if that is an issue, since you have already commited the changes to the lock file.

jakovljevic-mladen commented:
> I don't know if that is an issue

It shouldn't be an issue. Doing an `npm install` changes `package-lock.json` almost every time, but we have the latest `package-lock.json`, so we should be fine. I'll merge as soon as we fix the _Node 14 build_ issue.

jakovljevic-mladen commented:
Thanks for the fix @tmair and sorry for waiting this long for the merge. It is cherry-picked to `7.x` branch and deployed to the [rxjs.dev](https://rxjs.dev/).
